modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/type/IgniteTypeFactory.java

```diff
@@ -315,7 +316,7 @@ public RelDataType createCustomType(Type type) {
     public RelDataType createCustomType(Type type, boolean nullable) {
         if (UUID.class == type)
             return canonize(new UuidType(nullable));
-        else if (Object.class == type)
+        else if (Object.class == type || (type instanceof Class && BinaryObject.class.isAssignableFrom((Class<?>)type)))
             return canonize(new OtherType(nullable));
```

Comment:
Actually, `canonize()` is not working for `IgniteCustomType`. Does nothing

Reply:
It canonized by `DATATYPE_CACHE.intern` and always return the same instance for `OtherType`. Why do you think it does nothing?

Reply:
This is the reason. It always return the same value.